docs/_data/nakadi-event-bus-api.yaml

```
@@ -2858,6 +2861,26 @@ definitions:
         description: |
           Number of milliseconds that Nakadi stores events published to this event type.
 
+
+  Discriminator:
+    type: object
+    description: |
+      Event field parameters to differentiate consumption of events published on per-consumer level.
+      The discriminator can be used to specify a field to be used from the events published to this event type.
+      This field can then be used to implement logic using a plugin to whether an event would be provided
+      in a consumer's stream (`EventsStreamBatch` or `SubscriptionEventStreamBatch`).
+      Once created, the discrimator definition for an event-type can't be updated or removed.
+    properties:
+      path:
+        type: string
+        description: |
+          Path of the field in the event that contains the identifiers of the entities that can access the event.
```

I don't think that this is the right approach.
Now the event producer needs to know the names of the services (or users) which are allowed to read each individual event? This makes the producer's job unnecessarily hard (it needs to be updated whenever a new consumer needs data) – and old data can't ever be read by new consumers –, and puts data into the event which does not belong there.

Maybe a better approach is that the event data contains some classifier (which is indicated by this property), which then can be mapped in the event type's authorization setting to a set of services/users.

For example, an order event could have a business partner ID, and each service then belongs to a business partner (or several of them) and can only read events for its own business partner ID.

This whole field could possibly then move to the authorization object (to make it obvious that the discrimiator's purpose is authorization). (Though you'll need to validate that the path points to an existing (potentially required) field in the schema, I guess?)

The authorization object could then look like this.

```yaml
  authorization:
     discriminator:
       path: 'data/partner_code'
     writers: [...] # as now
     admins: [...] # as now
     readers:
        # those can read everything
        - data_type: service
          value: order_archiver
        # those can read just the indicated class
        - data_type: service
          value: partner1-order-mapper
          discriminator_value: partner1
        - data_type: team
          value: team-partner1-support
          discriminator_value: partner1
```
(I guess the exact syntax in the `readers` section would depend on the authorization plugin – I can even imagine having some logical expression here, or require some OAuth scope.)

For alternative names: `classifier` would be another option.

---

After reading the (Zalando-internal) design document, it looks like you already want to do something like this (possibly with the authorization also based on the OAuth token – that is part of the authorization plugin).

Then it just needs better documentation. 

```suggestion
          Path of the field in the event that contains a classifier, which will be used by an authorization plugin (together with definitions in the authorization section) to define who can read the event.
```

---

Agreed on all points. I've tried and updated the documentation to correctly reflect the purpose of the field as you suggested. We also decided to name the field as `event_auth_field` (aligning with other field specifiers like `ordering_key_fields`, etc). However, we don't make it part of authorization section as mainly, we want to keep the field immutable once defined.